src/script/sign.h

```diff
@@ -497,9 +498,35 @@ struct PartiallySignedTransaction
     void SanitizeForSerialization();
 };
 
+class PSBTSignatureCreator : public MutableTransactionSignatureCreator {
+private:
+    std::map<CPubKey, std::vector<unsigned char>>* sigs;
+    bool finalize_only;
+
+public:
+    PSBTSignatureCreator(std::map<CPubKey, std::vector<unsigned char>>* sigs, CMutableTransaction* tx, unsigned int nInIn, const CAmount& amountIn, int nHashTypeIn) : MutableTransactionSignatureCreator(tx, nInIn, amountIn, nHashTypeIn), sigs(sigs) {}
+    bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& address, const CScript& scriptCode, SigVersion sigversion) const override;
+};
+
+class PSBTSigningProvider : public SigningProvider {
+private:
+    std::vector<SigningProvider*> providers;
+    PartiallySignedTransaction* psbt;
+
+public:
+    bool GetCScript(const CScriptID &scriptid, CScript& script) const override;
+    bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const override;
+    bool GetKey(const CKeyID &address, CKey& key) const override;
+    void AddSigningProvider(SigningProvider* provider);
+
+    PSBTSigningProvider(PartiallySignedTransaction* psbt) : psbt(psbt) {}
+};
+
 /** Produce a script signature using a generic signature creator. */
 bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreator& creator, const CScript& scriptPubKey, SignatureData& sigdata);
 
+bool SignPartiallySignedTransaction(PartiallySignedTransaction& psbt, SigningProvider* provider, int nHashType, bool finalize = false);
```

Comment:
In commit "Implement a signer that takes an unserialized PSBT and signs it":

Maybe add a comment about this function, in particular saying that it returns true if it was complete, and if `finalize` was set, and `false` does not necessarily mean failure to partially sign.

Reply:
Done